apps/designer/app/designer/shared/tree/tree.tsx

```diff
@@ -1,132 +1,670 @@
+
+  // onBlur doesn't fire when the activeElement is removed from the DOM
+  useEffect(() => {
+    const haveFocus =
+      rootRef.current?.contains(document.activeElement) === true;
+    hadFocus.current = haveFocus;
+  });
+
+  return {
+    rootRef(element: HTMLElement | null) {
+      rootRef.current = element;
+    },
+    handleClick(event: React.MouseEvent<Element>) {
+      // When clicking on an item button make sure it gets focused.
+      // (see https://zellwk.com/blog/inconsistent-button-behavior/)
+      const itemButton = (event.target as HTMLElement).closest(
+        "[data-item-button-id]"
+      );
+      if (itemButton instanceof HTMLElement) {
+        itemButton.focus();
+        return;
+      }
+
+      // When clicking anywhere else in the tree,
+      // make sure the selected item doesn't loose focus.
+      if (selectedInstanceId !== undefined) {
+        setFocus(selectedInstanceId);
+      }
+    },
+    handleBlur() {
+      hadFocus.current = false;
+    },
+  };
+};
+
+const useExpandState = ({
+  selectedInstanceId,
+  root,
+}: {
+  root: Instance;
+  selectedInstanceId?: Instance["id"];
+}) => {
+  const [record, setRecord] = useState<Record<Instance["id"], boolean>>({});
+
+  // We want to automatically expand all parents of the selected instance whenever it changes
+  const prevSelectedInstanceId = useRef(selectedInstanceId);
+  const prevRoot = useRef(root);
+  useEffect(() => {
+    if (
+      selectedInstanceId === prevSelectedInstanceId.current &&
+      prevRoot.current === root
+    ) {
+      return;
+    }
+    prevSelectedInstanceId.current = selectedInstanceId;
+    prevRoot.current = root;
+    if (selectedInstanceId === undefined) {
+      return;
+    }
+    const path = getInstancePath(root, selectedInstanceId).map(
+      (instance) => instance.id
+    );
+    // Don't want to expand the selected instance itself
+    path.pop();
+    const toExpand = path.filter((id) => record[id] !== true);
+    if (toExpand.length === 0) {
+      return;
+    }
+    setRecord((record) => {
+      const newRecord = { ...record };
+      for (const id of toExpand) {
+        newRecord[id] = true;
+      }
+      return newRecord;
+    });
+  }, [record, root, selectedInstanceId]);
+
+  const getIsExpanded = useCallback(
+    (instance: Instance) => {
+      // root is always expanded
+      if (instance.id === root.id) {
+        return true;
+      }
+
+      return getIsExpandable(instance) && record[instance.id] === true;
+    },
+    [record, root]
+  );
+
+  const setIsExpanded = useCallback((instance: Instance, expanded: boolean) => {
+    setRecord((record) => ({ ...record, [instance.id]: expanded }));
+  }, []);
+
+  return { getIsExpanded, setIsExpanded };
+};
+
+type ShiftedDropTarget = {
+  instance: Instance;
+  position: number | "end";
+  placement:
+    | { type: "rect"; rect: Rect }
+    | { type: "line"; placement: Placement };
+};
+
+export const useHorizontalShift = ({
```

Comment:
this export isn't used outisde of tree, but is exported, I think you want to move this hook into a separate file and make sure that you only export the things that are actually public. Look at index.ts files like they represent the public interface, please check all index files in this PR